core/Enclave/enclave_hsm.cpp

```diff
@@ -42,768 +41,699 @@
-        return UINT32_MAX;
-
-    if (plaintext_size > UINT32_MAX - sizeof(sgx_aes_gcm_data_ex_t))
-        return UINT32_MAX;
-
-    if (aad_size > UINT32_MAX - plaintext_size)
-        return UINT32_MAX;
-
-    if (sizeof(sgx_aes_gcm_data_ex_t) > UINT32_MAX - plaintext_size - aad_size)
-        return UINT32_MAX;
-
-    return (aad_size + plaintext_size + sizeof(sgx_aes_gcm_data_ex_t));
-}
-
-static uint32_t sgx_get_gcm_ciphertext_size(const sgx_aes_gcm_data_ex_t *gcm_data)
+static uint32_t get_asymmetric_max_encrypt_plaintext_size(ehsm_keyspec_t keyspec, ehsm_padding_mode_t padding)
 {
-    if (NULL == gcm_data)
-        return UINT32_MAX;
-
-    return gcm_data->ciphertext_size;
-}
-
-
-static sgx_status_t sgx_gcm_encrypt(const sgx_aes_gcm_128bit_key_t *key,
-                                    const uint32_t plaintext_size, const uint8_t *plaintext,
-                                    const uint32_t aad_size, const uint8_t *aad,
-                                    const uint32_t gcm_data_size, sgx_aes_gcm_data_ex_t *gcm_data)
-{
-    if (NULL == gcm_data || gcm_data_size < sgx_calc_gcm_data_size(aad_size, plaintext_size))
-        return SGX_ERROR_INVALID_PARAMETER;
-
-    if (plaintext_size == 0 || NULL == plaintext)
-        return SGX_ERROR_INVALID_PARAMETER;
-
-    if (plaintext_size > UINT32_MAX - aad_size)
-        return SGX_ERROR_INVALID_PARAMETER;
-
-    uint32_t real_aad_size = aad_size;
-    if (NULL == aad)
-        real_aad_size = 0;
-
-    sgx_status_t ret = sgx_read_rand(gcm_data->iv, sizeof(gcm_data->iv));
-    if (ret != SGX_SUCCESS) {
-        printf("error generating iv.\n");
-        return ret;
-    }
-
-    ret = sgx_rijndael128GCM_encrypt(key,
-                                     plaintext, plaintext_size,
-                                     gcm_data->payload,
-                                     gcm_data->iv, sizeof(gcm_data->iv),
-                                     &(gcm_data->payload[gcm_data->ciphertext_size]), real_aad_size,
-                                     reinterpret_cast<uint8_t (*)[16]>(gcm_data->mac));
-    if (SGX_SUCCESS != ret) {
-        printf("gcm encrypting failed.\n");
-    }
-    else {
-        gcm_data->ciphertext_size = plaintext_size;
-        gcm_data->aad_size = real_aad_size;
+    uint32_t padding_size;
+    switch (padding)
+    {
+    case EH_PAD_RSA_PKCS1:
+        padding_size = 11;
+        break;
+    case EH_PAD_RSA_PKCS1_OAEP:
+        padding_size = 42; // where is 42 from: https://github.com/openssl/openssl/blob/master/crypto/rsa/rsa_oaep.c
+        break;
+    case EH_PAD_RSA_NO:
+    default:
+        padding_size = 0;
+        break;
+    }
+    switch (keyspec)
+    {
+    case EH_RSA_2048:
+        return 256 - padding_size;
+    case EH_RSA_3072:
+        return 384 - padding_size;
+    case EH_RSA_4096:
+        return 512 - padding_size;
+    case EH_SM2:
+        return 64; // why 64: sm2 key length is 256 bits
```

Comment:
why 256-bit key length return 64 bit plaintext_size?

Reply:
what's the length of public key and private key of SM2?

Reply:
> why 256-bit key length return 64 bit plaintext_size?

here has some problem, I'll check later



> what's the length of public key and private key of SM2?

private:32, public:64 byte